chef/cookbooks/nova/recipes/ceph.rb

```diff
@@ -0,0 +1,187 @@
+      Chef::Log.info("Ceph cluster is not healthy; skipping the ceph setup for backend #{volume[:backend_name]}")
+      next
+    end
+  end
+
+  rbd_user = volume[:rbd][:user]
+  rbd_uuid = volume[:rbd][:secret_uuid]
+
+  secret_file_path = "/etc/ceph/ceph-secret-#{rbd_uuid}.xml"
+  
+  file secret_file_path do
+    owner "root"
+    group "root"
+    mode "0640"
+    content "<secret ephemeral='no' private='no'> <uuid>#{rbd_uuid}</uuid><usage type='ceph'> <name>client.#{rbd_user} secret</name> </usage> </secret>"
+  end #file secret_file_path
+
+  ruby_block "save nova key as libvirt secret" do
+    block do
+      # Check if libvirt was installed
+      if system("virsh hostname &> /dev/null")
+
+        # First remove conflicting secrets due to same usage name
+        virsh_secret = Mixlib::ShellOut.new("virsh secret-list")
+        secret_list = virsh_secret.run_command.stdout
+        virsh_secret.error!
+
+        secret_lines = secret_list.strip.split("\n")
+        if secret_lines.length < 2 || !secret_lines[0].start_with?("UUID") || !secret_lines[1].start_with?("----")
+          raise "cannot fetch list of libvirt secret"
+        end
+        secret_lines.shift(2)
+
+        secret_lines.each do |secret_line|
+          secret_uuid = secret_line.split(" ")[0]
+          cmd = ["virsh", "secret-dumpxml", secret_uuid]
+          virsh_secret_dumpxml = Mixlib::ShellOut.new(cmd)
+          secret_xml = virsh_secret_dumpxml.run_command.stdout
+          virsh_secret_dumpxml.error!
+          # some secrets might not be ceph-related, skip these
+          next if secret_xml.index("<usage type='ceph'>").nil?
+
+          # lazy xml parsing
+          re_match = %r[<usage type='ceph'>.*<name>(.*)</name>]m.match(secret_xml)
+          next if re_match.nil?
+          secret_usage = re_match[1]
+          undefine = false
+
+          if secret_uuid == rbd_uuid
+            undefine = true if secret_usage != "client.#{rbd_user} secret"
+          else
+            undefine = true if secret_usage == "client.#{rbd_user} secret"
+          end
+
+          if undefine
+            cmd = ["virsh", "secret-undefine", secret_uuid]
+            virsh_secret_undefine = Mixlib::ShellOut.new(cmd)
+            virsh_secret_undefine.run_command
+            virsh_secret_undefine.error!
+          end
+        end
+
+        # Now add our secret and its value
+        cmd = ["ceph", "-k", admin_keyring, "-c", ceph_conf, "-s", "auth", "get-key", "client.#{rbd_user}" ]
+        ceph_get_key = Mixlib::ShellOut.new(cmd)
+        client_key = ceph_get_key.run_command.stdout
+        ceph_get_key.error!
+
+        cmd = ["virsh", "secret-get-value", rbd_uuid ]
+        virsh_secret_get_value = Mixlib::ShellOut.new(cmd)
+        secret = virsh_secret_get_value.run_command.stdout.chomp.strip
+
+        if secret != client_key
+          # It seems that virsh secret-define returns '' empty string
+          # instead of exitcodes, like [0] and etc.
```

Comment:
Thanks for finishin the comment :-) That being said, I'm not totally clear what it means and if it explains something that is done in the code or something else. Maybe I'm missing some context?